external-crates/move/crates/move-package/tests/test_lock_file.rs

```diff
@@ -101,6 +103,66 @@ fn update_lock_file_toolchain_version() {
     expected.assert_eq(&toml);
 }
 
+#[test]
+fn test_update_managed_address() {
+    let pkg = create_test_package().unwrap();
+    let lock_path = pkg.path().join("Move.lock");
+
+    // Initialize lock file.
+    let lock = LockFile::new(
+        pkg.path().to_path_buf(),
+        /* manifest_digest */ "42".to_string(),
+        /* deps_digest */ "7".to_string(),
+    )
+    .unwrap();
+    lock.commit(&lock_path).unwrap();
+
+    // Update managed address in lock file.
+    let pb = PathBuf::from(pkg.path());
+    let mut lock = LockFile::from(pb, &lock_path).unwrap();
+    update_managed_address(
+        &mut lock,
+        /* chain_identifier */ "35834a8a".into(),
+        ManagedAddressUpdate::Published {
+            original_id: "0x123".into(),
+        },
+    )
+    .unwrap();
+
+    update_managed_address(
+        &mut lock,
+        /* chain_identifier */ "35834a8a".into(),
+        ManagedAddressUpdate::Upgraded {
+            latest_id: "0x456".into(),
+            version: 2,
+        },
+    )
+    .unwrap();
+    lock.commit(&lock_path).unwrap();
+
+    // Read lock file and check contents.
+    let mut lock_file = File::open(lock_path).unwrap();
+    let contents = {
+        let mut buf = String::new();
+        let _ = lock_file.read_to_string(&mut buf);
+        buf
+    };
+
+    let expected = expect![[r##"
+        # @generated by Move, please check-in and do not edit manually.
+
+        [move]
+        version = 0
+        manifest_digest = "42"
+        deps_digest = "7"
+
+        ["move.env.35834a8a.managed-addresses"]
+        original-id = "0x123"
+        latest-id = "0x456"
+        version = "2"
```

Comment:
Want to specially call out the format and decision behind `["move.env.35834a8a.managed-addresses"]` here:

- choosing to create an independent `env` index specifically so that it's easy to collect / deserialize `env` entries once there exist multiple chain IDs

- choosing to put the chain ID right after `env`, instead of a postfix like `env.managed-addresses.35834a8a`. This so that we can namespace other things like `[[move.package]]` under the chain-identifier prefix in follow up. In other words, we'll have:

```toml
[["move.env.35834a8a.packages"]]
```

- For `env` sections, I've prefixed it with the `move`: I do not feel strongly about this, and I know @amnn hesitated on this. We can remove it and just have the section start with `env`. For consistency with  `[[move.package]]` I thought it made sense to use `move`. Now's the time to speak up if we don't want to roll with the `move` prefix!

Reply:
Related to https://github.com/MystenLabs/sui/pull/14178#issuecomment-1924760426:

I suspect we'll be able to do something like

```toml
[["move.env.35834a8a"]]
 # IDs for package when resolved during a build, in support of on-chain dependencies
root = {
  id = "0x123" # corresponds to `original-id` if exists, or name from manifest
  version = "0x456" # Corresponds to `latest-id`, None if it doesn't exist
}

[["move.env.package.35834a8a"]]
# ...
```

However: enabling this would need a way to establish `env` (i.e., chain identifier) of a user if they only build, but do not necessarily publish. Which is a bit odd because they may not publish to that chain. We are able to populate and persist this info once they `sui client publish` / `sui client upgrade` because we understand the intent, but not necessarily before. 

We can try resolve chain ID based on current CLI env (rpc), the same way as if the user would run `sui client publish` at that moment. It still leaves open the possibility that nothing exists, or, say the rpc client is configured to `localnet` that may not be running.

Open thought is we can have a kind of `move.env.staging` with a predictable name for the build case (this content may change the lock file with different build info if the user changes their env).

Reply:
(To reviewers: On this last comment, I don't think the outcome of discussion is blocking for this PR)

Reply:
Hmm, good question. I think this is a more general problem of "there's no env" (which can also happen e.g. when the cli context is not configured to any rpc). One thing to consider is how to resolve the package id and version in that case because with the hooks we want to resolve them to the original-id and latest-id / version but we can't do that without knowing the chain id / env. So once you answer that the answer to your question will become more clear. 

I think that the chain id / env should always be set or default to mainnet. Also I feel like there should somehow be more orthogonality between RPC and chain / env. Like the workflow should be that you're switching between envs (instead of RPCs) and we can have the `mainnet` and `testnet` envs be pre-populated by default (with their chain ids). Then you're able to set RPCs for each env specifically, have a default one, and switch between them (where you can error / warn on setting wrong RPC for env based on chain-id check). This way you don't have to have a RPC set to have a set env and we can have the `mainnet` env be default.

Reply:
> Want to specially call out the format and decision behind `["move.env.35834a8a.managed-addresses"]` here:
> 
> * choosing to create an independent `env` index specifically so that it's easy to collect / deserialize `env` entries once there exist multiple chain IDs
> * choosing to put the chain ID right after `env`, instead of a postfix like `env.managed-addresses.35834a8a`. This so that we can namespace other things like `[[move.package]]` under the chain-identifier prefix in follow up. In other words, we'll have:
> 
> ```toml
> [["move.env.35834a8a.packages"]]
> ```
> 
> * For `env` sections, I've prefixed it with the `move`: I do not feel strongly about this, and I know @amnn hesitated on this. We can remove it and just have the section start with `env`. For consistency with  `[[move.package]]` I thought it made sense to use `move`. Now's the time to speak up if we don't want to roll with the `move` prefix!

I vote for no "move" prefix. The file is already called `Move.lock`, we know it's Move, no need to restate it in the section name.

Reply:
noted, thanks @kklas!

Reply:
> I think that the chain id / env should always be set or default to mainnet

I like the idea of a default, although if it has to be tied to a network, I would prefer testnet, given that it makes it harder to make a mistake that has monetary consequences. I think that can be avoided entirely though (see below).

Regarding the problem of what happens before publish/upgrade. I'm thinking along the lines of the following constraints (and agreed that this conversation mostly does not block this PR, but it may have small implications on e.g. the chain identifier portion of the lock file format):

- We need to have a notion of environment that's decoupled from the RPC/chain (as per @kklas' comment), because 
  - otherwise the story gets murky before the first time the package is published (and also in the case of source-only packages that are never published by themselves), and
  - the idea of having multiple environments per chain is in line with our broader motivation for envs (one could imagine a single source file published multiple times to the same chain).
- As much as possible, we should avoid having developers need to mess around with chain identifiers, because they are confusing -- human-readable names all the way.
- Most people should not have to worry about different environments, especially not at first, it's somewhat of a power user feature and so people's exposure to the associated complexity (e.g. in the manifest) should be proportional to how complicated their aims are.
- The environment in one package can select alternate environments in its dependencies.

This points me to generally a set-up where environments are present in the manifest as a sort of "overlay": 

- Environments can control the presence or absence of dependencies, the environment for those dependencies, and the assignment of addresses.
  - This then starts to look like an extension of the current `dev` and `test` modes.
- Chain identifiers are optional in environments, but if they are provided, they need to be consistent.
- There is a "default" or "base" environment that is defined by the contents of the manifest as it appears today.
- When writing out the manifest as part of `sui move new`, we write out some environments for people, based on their active environments.
- Dependencies can additionally have an environment override that says that we are using that dependency specialised to one of its environments, rather than the default.
  - We could also (later) support marking dependencies as "optional" (which means they are not included by default), but this would require adding source language support for guarding access to these dependencies by the environment.
  - It may be worth supporting propagating environments (so, by default, if you are compiling package foo with environment "mainnet" then it propagates that environment to its dependencies by default, unless its own configuration for that dependency under that environment says differently).

Putting it all together, the manifest ends up looking something like this:

```toml
[move]
name = "Foo"

[dependencies]
Bar = { local = "..." }
Baz = { git = "...", rev = "...", env = "special" } # override Baz's environment by default
Qux = { id = "..." }

[env.dev.dependencies]
Bar.env = "default" # build Bar in normal mode even when everything else is in dev mode.
Baz.rev = "dev" # Even override the core parts of the dependency (like where it is found)

[env.mainnet]
chain-identifier = "35834a8a" # Generated for you by `sui move new`

[env.mainnet.dependencies]
Qux = { id = "..." } # Override the on-chain dependency for mainnet, assuming default is testnet

```

P.S. Note the quotes around the TOML key headers in this test, that's the smoking gun that indicates to me that this TOML file doesn't actually have nested keys, it has a single key that happens to contain dots in it, as per my earlier comment.

Reply:
This makes sense to me in general. Though I think that the structure of this new manifest format with env support should be more precisely described and discussed before committing to anything.

Another thing to consider w.r.t. the dependency resolution with env overrides and multiple envs for the same chain is that if we resolve the package identifier to `original-id` as planed, it could cause consistency issues across the graph when you have the same dependency referenced somewhere else transitively but with a different chain-id. This will make the resolver consider these as two separate packages and ultimately cause duplicate definition errors once it goes into compilation. So these env overrides will probably need to work like the regular overrides where all instances of the dep are replaced throughout the graph, but it's not exactly clear to me how this will be implemented since regular overrides work on overriding packages of the same identifier (`original-id`) where the identifier stays the same after the override and here you'll have to be switching out the identifier also with a different one. So the dependency graph will likely need more tweaking to support this.

Reply:
> Though I think that the structure of this new manifest format with env support should be more precisely described and discussed before committing to anything.

Yes agreed, this needs to be designed properly before we build anything, and what I've shared is just a very rough sketch.

> Another thing to consider w.r.t. the dependency resolution with env overrides and multiple envs for the same chain is that if we resolve the package identifier to original-id as planed, it could cause consistency issues across the graph

My mental model for this is that we are creating `N` different dependency graphs, one for each environment, and the requirements around consistency apply only within each environment, not across all of them. The fact that environments are defined as deltas on-top of a default is purely a convenience at the source level.